core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -75,197 +105,658 @@ object compress {
-    * which throws exceptions when exhausted rather than blocking. These signal
-    * exceptions are caught by the pipe and treated as an async suspension. Thus,
-    * there are no issues with arbitrarily-framed data and chunk boundaries. Also
-    * note that there is almost no performance impact from these exceptions, due
-    * to the way that the JVM handles throw/catch.
-    *
-    * The chunk size here is actually really important. If you set it to be too
-    * small, then there will be insufficient buffer space for `GZIPInputStream` to
-    * read the GZIP header preamble. This can result in repeated, non-progressing
-    * async suspensions. This case is caught internally and will be raised as an
-    * exception (`NonProgressiveDecompressionException`) within the output stream.
-    * Under normal circumstances, you shouldn't have to worry about this. Just, uh,
-    * don't set the buffer size to something tiny. Matching the input stream largest
-    * chunk size, or roughly 8 KB (whichever is larger) is a good rule of thumb.
+    * Returns a `Pipe` that gunzips (decompresses) its input elements as defined by RFC 1952:
+    * https://www.ietf.org/rfc/rfc1952.txt
     *
-    * @param bufferSize The bounding size of the input buffer. This should roughly
-    *                   match the size of the largest chunk in the input stream.
-    *                   The chunk size in the output stream will be determined by
-    *                   double this value.
+    * @param bufferSize size of the internal buffer that is used by the decompressor. Default size is 32 KB.
     */
   def gunzip[F[_]: RaiseThrowable](bufferSize: Int): Pipe[F, Byte, Byte] =
-    in =>
-      Stream.suspend {
-        val abis: AsyncByteArrayInputStream = new AsyncByteArrayInputStream(bufferSize)
+    gunzip(bufferSize, _ => (), _ => ())
 
-        def push(chunk: Chunk[Byte]): Unit = {
-          val arr: Array[Byte] = {
-            val buf = new Array[Byte](chunk.size)
-            chunk.copyToArray(buf) // Note: we can be slightly better than this for Chunk.Bytes if we track incoming offsets in abis
-            buf
+  /**
+    * Returns a `Pipe` that gunzips (decompresses) its input elements as defined by RFC 1952:
+    * https://www.ietf.org/rfc/rfc1952.txt
+    *
+    * @param bufferSize size of the internal buffer that is used by the decompressor. Default size is 32 KB.
+    * @param modificationTimeCallback callback to provide the optional file modificationt time, if included in the gzip stream
+    * @param fileNameCallback callback to provide the optional file name, if included in the gzip stream.
+    * @param commentCallback callback to provide the optional file comment, if included in the gzip stream.
+    */
+  def gunzip[F[_]: RaiseThrowable](
+      bufferSize: Int = 1024 * 32,
+      modificationTimeCallback: Instant => Unit = _ => (),
+      fileNameCallback: String => Unit = _ => (),
+      commentCallback: String => Unit = _ => ()
```

Comment:
In general, I think callbacks like this are not great. At the very least, they should return `F[Unit]` so that effects can be properly controlled and sequenced. But aside from that, I would definitely prefer if there were a way to sequence this into the stream more properly.

Reply:
Agreed, I don't love this callback approach either, but don't have a solution for this yet.

Regarding wrapping the the header property values in F, I'm not sure I know how to do this. F is currently only a RaiseThrowable, how F.unit or F.pure be used? There is a similar problem preventing the Inflater and Deflater instances from being bracketed. Is it possible to further constrain F while maintaining compatibility - what am I missing?

Regarding an alternative to callback parameters and sequencing the stream, what do you think about removing the callback parameters and changing the return type to return a stream of a single tuple (or an appropriate structure) that would return the materialised header properties and the remaining stream? I'd have to tinker with it, but this should be possible, unless there is a better idea. An additional benefit of this approach would be that it would be possible to discard the compressed content based on the content of the header properties.

Reply:
> F is currently only a RaiseThrowable, how F.unit or F.pure be used?

I think you could go with a "partially applied" method, so this one wouldn't take callbacks (it would take `Applicative[F]`) but would create an instance of a class that has an `apply` that takes callbacks (in there, you could use the Applicative instance in default parameters).

Reply:
:+1: for returning a case class with the header properties and the body stream.

Reply:
> F is currently only a RaiseThrowable, how F.unit or F.pure be used

`F` is `RaiseThrowable` *here*, but the call site probably knows something more about it! For example, the call site may know that `F` is actually `IO`, which would allow it to wrap side effects into `F[Unit]` values. But either way the case class idea is way better.

Reply:
You're right, thanks. I'm going to replace the callback parameters with a case class return...

Reply:
Right, callback parameters removed from fs2.compress2.gunzip and replaced with a returned case class containing the materialised header fields and a body stream. Opinions on naming welcome.